File: sdk/python/examples/langgraph/17_error_recovery.py
# Copyright (c) 2025 Agentspan
# Licensed under the MIT License. See LICENSE file in the project root for details.

"""Error Recovery — create_agent with graceful error handling in tools.

Demonstrates:
    - Catching exceptions within @tool functions for graceful degradation
    - Returning error information as a string so the LLM can handle it
    - The agent decides whether to process data or recover from the error

Requirements:
    - AGENTSPAN_SERVER_URL=http://localhost:8080/api
    - OPENAI_API_KEY for ChatOpenAI
"""

from langchain_core.tools import tool
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI
from agentspan.agents.langchain import create_agent
from agentspan.agents import AgentRuntime

llm = ChatOpenAI(model="gpt-4o-mini", temperature=0)


@tool
def fetch_data(query: str) -> str:
    """Attempt to fetch data for a query. May fail for certain query patterns.

    Returns fetched data on success, or an error description if the fetch fails.
    Queries containing 'fail' or 'error' will simulate a transient failure.

    Args:
        query: The data query to fetch.
    """
    try:
        if "fail" in query.lower() or "error" in query.lower():
            raise ValueError(f"Simulated fetch failure for query: '{query}'")
        return (
            f"Fetched data for '{query}': "
            "Sample dataset with 100 records, avg value 42.5, max 99, min 1."
        )
    except Exception as exc:
        return f"ERROR: {exc}"


@tool
def analyze_data(data: str) -> str:
    """Summarize fetched data in one sentence.

    Args:
        data: The fetched data string to analyze.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", "You are a data analyst. Summarize the following data in one sentence."),
        ("human", "{data}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"data": data})
    return response.content


@tool
def suggest_recovery(error_message: str, original_query: str) -> str:
    """Generate a helpful recovery message when a data fetch has failed.

    Apologizes briefly, explains what may have gone wrong, and suggests
    2 alternative approaches.

    Args:
        error_message: The error that occurred.
        original_query: The original query that failed.
    """
    prompt = ChatPromptTemplate.from_messages([
        ("system", (
            "A data fetch error occurred. Apologize briefly, explain what may have gone wrong, "
            "and suggest 2 alternative approaches the user could try. Be concise."
        )),
        ("human", "Error: {error}\nOriginal query: {query}"),
    ])
    chain = prompt | llm
    response = chain.invoke({"error": error_message, "query": original_query})
    return f"[RECOVERED FROM ERROR]\n{response.content}"


RECOVERY_SYSTEM = """You are a resilient data assistant.

For each query:
1. Call fetch_data to retrieve data
2. Check the result:
   - If the result starts with 'ERROR:' → call suggest_recovery with the error and original query
   - Otherwise → call analyze_data to summarize the successful fetch
3. Return the final result to the user
"""

graph = create_agent(
    llm,
    tools=[fetch_data, analyze_data, suggest_recovery],
    name="error_recovery_agent",
    system_prompt=RECOVERY_SYSTEM,
)

if __name__ == "__main__":
    with AgentRuntime() as runtime:
        print("=== Happy path ===")
        result = runtime.run(graph, "sales data for Q4")
        print(f"Status: {result.status}")
        result.print_result()

        print("\n=== Error recovery path ===")
        result = runtime.run(graph, "intentionally fail this query")
        print(f"Status: {result.status}")
        result.print_result()

File: sdk/typescript/examples/langgraph/17-error-recovery.ts
/**
 * Error Recovery -- StateGraph with try/catch in nodes for graceful degradation.
 *
 * Demonstrates:
 *   - Catching exceptions within StateGraph nodes
 *   - Storing error information in state for downstream handling
 *   - A fallback node that generates a graceful response on failure
 *   - Conditional routing based on whether an error occurred
 */

import { StateGraph, START, END, Annotation } from '@langchain/langgraph';
import { ChatOpenAI } from '@langchain/openai';
import { HumanMessage, SystemMessage } from '@langchain/core/messages';
import { AgentRuntime } from '../../src/index.js';

const llm = new ChatOpenAI({ model: 'gpt-4o-mini', temperature: 0 });

// ---------------------------------------------------------------------------
// State schema
// ---------------------------------------------------------------------------
const RecoveryState = Annotation.Root({
  query: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  data: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  error: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
  response: Annotation<string>({
    reducer: (_prev: string, next: string) => next ?? _prev,
    default: () => '',
  }),
});

type State = typeof RecoveryState.State;

// ---------------------------------------------------------------------------
// Node functions
// ---------------------------------------------------------------------------
function fetchData(state: State): Partial<State> {
  const query = state.query;
  try {
    // Simulate a failure for queries containing 'fail' or 'error'
    if (query.toLowerCase().includes('fail') || query.toLowerCase().includes('error')) {
      throw new Error(`Simulated fetch failure for query: '${query}'`);
    }

    // Simulate successful data fetch
    const data =
      `Fetched data for '${query}': ` +
      'Sample dataset with 100 records, avg value 42.5, max 99, min 1.';
    return { data, error: '' };
  } catch (exc: any) {
    // Capture the error in state instead of crashing the graph
    return { data: '', error: String(exc.message ?? exc) };
  }
}

function shouldRecover(state: State): string {
  return state.error ? 'recover' : 'process';
}

async function processData(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'You are a data analyst. Summarize the following data in one sentence.',
    ),
    new HumanMessage(state.data),
  ]);
  return { response: response.content as string };
}

async function recoverFromError(state: State): Promise<Partial<State>> {
  const response = await llm.invoke([
    new SystemMessage(
      'A data fetch error occurred. Apologize briefly, explain what may have gone wrong, ' +
        'and suggest 2 alternative approaches the user could try. Be concise.',
    ),
    new HumanMessage(`Error: ${state.error}\nOriginal query: ${state.query}`),
  ]);
  return { response: `[RECOVERED FROM ERROR]\n${response.content}` };
}

// ---------------------------------------------------------------------------
// Build the graph
// ---------------------------------------------------------------------------
const builder = new StateGraph(RecoveryState);
builder.addNode('fetch', fetchData);
builder.addNode('process', processData);
builder.addNode('recover', recoverFromError);

builder.addEdge(START, 'fetch');
builder.addConditionalEdges('fetch', shouldRecover, {
  process: 'process',
  recover: 'recover',
});
builder.addEdge('process', END);
builder.addEdge('recover', END);

const graph = builder.compile();

// Add agentspan metadata for extraction
(graph as any)._agentspan = {
  model: 'openai/gpt-4o-mini',
  tools: [],
  framework: 'langgraph',
};

// ---------------------------------------------------------------------------
// Run on agentspan
// ---------------------------------------------------------------------------
async function main() {
  const runtime = new AgentRuntime();
  try {
    // Deploy to server. CLI alternative (recommended for CI/CD):
    //   agentspan deploy <module>
    // await runtime.deploy(graph);
    // await runtime.serve(graph);
    // Direct run for local development:
    console.log('=== Happy path ===');
    let result = await runtime.run(graph, 'sales data for Q4');
    console.log('Status:', result.status);
    result.printResult();

    console.log('\n=== Error recovery path ===');
    result = await runtime.run(graph, 'intentionally fail this query');
    console.log('Status:', result.status);
    result.printResult();
  } finally {
    await runtime.shutdown();
  }
}

// Only run when executed directly (not when imported for discovery)
if (process.argv[1]?.endsWith('17-error-recovery.ts') || process.argv[1]?.endsWith('17-error-recovery.js')) {
  main().catch(console.error);
}
